action.yml

```diff
@@ -30,77 +56,77 @@ runs:
       uses: actions/download-artifact@v2
       with:
         name: nodejs-sdk.tar.gz
         path: ${{ github.workspace }}/sdk/
     - name: Uncompress nodejs SDK
-      if: inputs.sdk == 'nodejs' || inputs.sdk == 'all'
+      if: (contains(inputs.sdk, 'nodejs') || (contains(inputs.sdk, 'all') && !contains(inputs.sdk, '!all'))) && !contains(inputs.sdk, '!nodejs')
       run: tar -zxf ${{ github.workspace }}/sdk/nodejs.tar.gz -C
         ${{ github.workspace }}/sdk/nodejs
       shell: bash
     - name: Run npm whoami
-      if: inputs.sdk == 'nodejs' || inputs.sdk == 'all'
+      if: (contains(inputs.sdk, 'nodejs') || (contains(inputs.sdk, 'all') && !contains(inputs.sdk, '!all'))) && !contains(inputs.sdk, '!nodejs')
       run: npm whoami
       shell: bash
       env:
         NODE_AUTH_TOKEN: ${{ env.NODE_AUTH_TOKEN }}
     - name: Publish Node
-      if: inputs.sdk == 'nodejs' || inputs.sdk == 'all'
+      if: (contains(inputs.sdk, 'nodejs') || (contains(inputs.sdk, 'all') && !contains(inputs.sdk, '!all'))) && !contains(inputs.sdk, '!nodejs')
       run: pulumi package publish-sdk nodejs --path ${{ github.workspace }}/sdk/nodejs/bin
       shell: bash
       env:
         NODE_AUTH_TOKEN: ${{ env.NODE_AUTH_TOKEN }}
     - name: Setup DotNet
-      if: inputs.sdk == 'dotnet' || inputs.sdk == 'all'
+      if: (contains(inputs.sdk, 'dotnet') || (contains(inputs.sdk, 'all') && !contains(inputs.sdk, '!all'))) && !contains(inputs.sdk, '!dotnet')
       uses: actions/setup-dotnet@v3
       with:
         dotnet-version: ${{ env.DOTNETVERSION }}
     - name: Download dotnet SDK
-      if: inputs.sdk == 'dotnet' || inputs.sdk == 'all'
+      if: (contains(inputs.sdk, 'dotnet') || (contains(inputs.sdk, 'all') && !contains(inputs.sdk, '!all'))) && !contains(inputs.sdk, '!dotnet')
       uses: actions/download-artifact@v2
       with:
         name: dotnet-sdk.tar.gz
         path: ${{ github.workspace }}/sdk/
     - name: Uncompress dotnet SDK
-      if: inputs.sdk == 'dotnet' || inputs.sdk == 'all'
+      if: (contains(inputs.sdk, 'dotnet') || (contains(inputs.sdk, 'all') && !contains(inputs.sdk, '!all'))) && !contains(inputs.sdk, '!dotnet')
       run: tar -zxf ${{ github.workspace }}/sdk/dotnet.tar.gz -C
         ${{ github.workspace }}/sdk/dotnet
       shell: bash
     - name: Publish dotnet SDK
-      if: inputs.sdk == 'dotnet' || inputs.sdk == 'all'
+      if: (contains(inputs.sdk, 'dotnet') || (contains(inputs.sdk, 'all') && !contains(inputs.sdk, '!all'))) && !contains(inputs.sdk, '!dotnet')
       run: find "${{ github.workspace }}/sdk/dotnet/bin/Debug/" -name 'Pulumi.*.nupkg'
         -exec dotnet nuget push -k "${NUGET_PUBLISH_KEY}" -s https://api.nuget.org/v3/index.json {} \;
       shell: bash
     - name: Setup Python
-      if: inputs.sdk == 'python' || inputs.sdk == 'all'
+      if: (contains(inputs.sdk, 'python') || (contains(inputs.sdk, 'all') && !contains(inputs.sdk, '!all'))) && !contains(inputs.sdk, '!python')
```

Comment:
🤯 


This is good to go but let's not make too many more changes like this on top of this one or else we'll end up needing a SAT solver to understand what's going on.  I fully appreciate that GHA have limited abstraction capacity so this may be needed...

Reply:
We don't need a SAT solver... but I did build this to verify the logic:

```python

def contains(haystack: str, needle: str) -> bool:
    return needle in haystack

class input:
    def __init__(self, sdk):
        self.sdk = sdk

def check(query: str, sdk: str) -> bool:
    inputs = input(sdk)
    return eval(query.replace("||", "or").replace("&&", "and").replace(" !", " not "))

check("(contains(inputs.sdk, 'nodejs') || (contains(inputs.sdk, 'all') && !contains(inputs.sdk, '!all'))) && !contains(inputs.sdk, '!nodejs')", "!all")
```

We've shot ourselves in the foot several times with arrays vs strings through composite actions. It seems like everything is stringly typed given enough transformations. I don't think we will need to make any more changes here. 🤞

I might be able to refactor so the check only needs to show up once per language.